compiler/src/iree/compiler/Codegen/LLVMGPU/Verifiers.cpp

```diff
@@ -4,225 +4,265 @@
-    // dimensions for remaining comparisons.
+  if (auto batchMatmulOp = dyn_cast<linalg::BatchMatmulOp>(op)) {
+    // Inspect the batch tile dimensions separately for batch. The batch tile
+    // dim should be strictly greater than 1 for parallelizable loops and 0
+    // for non-parallelizable.
     if (cast<PartitionableLoopsInterface>(op).getPartitionableLoops(
-            kNumMaxParallelDims)[0] == 0) {  // The first dimension is
-      if (firstLevelTileSizes[0] > 1) {
-        return op->emitError("Received first tile dimension of ")
-               << firstLevelTileSizes[0] << " instead of 1 or 0 for "
+            kNumMaxParallelDims)[0] == 0) {
+      if (tileShape[0] > 1) {
+        return op->emitError("Received batch tile dimension of ")
+               << tileShape[0]
+               << " instead of 1 or lower for partitionable loops with "
                << pipelineName;
       }
     } else {
-      if (firstLevelTileSizes[0] != 0) {
-        return op->emitError("Received first tile dimension of ")
-               << firstLevelTileSizes[0] << " instead of 0 for "
+      if (tileShape[0] != 0) {
+        return op->emitError("Received batch tile dimension of ")
+               << tileShape[0]
+               << " instead of 0 for non-partitionable loops with"
                << pipelineName;
       }
     }
-    firstLevelTileSizes = {firstLevelTileSizes[1], firstLevelTileSizes[2],
-                           firstLevelTileSizes[3]};
-  }
 
-  // Verify the total workgroup size is <= 1024
-  int64_t totalWorkgroupSize =
-      workgroupSize[0] * workgroupSize[1] * workgroupSize[2];
-  if (totalWorkgroupSize > 1024) {
-    return op->emitOpError("expected workgroup size to be <=1024 for ")
-           << pipelineName << ", got " << totalWorkgroupSize;
+    // Remove the batch dimension from the threadBlockShape, lhsShape, and
+    // rhsShape.
+    threadBlockShape = {tileShape[1], tileShape[2], tileShape[3]};
+    lhsShape = lhsShape.drop_front();
+    rhsShape = rhsShape.drop_front();
   }
 
-  // Verify that the workgroup X dimension is 32 aligned
-  if (workgroupSize[0] % 32 != 0) {
-    return op->emitOpError("workgroup size is not 32 aligned for ")
-           << pipelineName << ", got " << workgroupSize[0];
-  }
-  if (softwarePipelinedepth > 1 && firstLevelTileSizes[2] == lhsShape[1]) {
-    return op->emitError(
-               "Software pipelining is not supported when first level K tile "
-               "size is same as matrix reduction size.\n This dispatch has\nk "
-               "tile: ")
-           << firstLevelTileSizes[2]
-           << "\nMatrix reduction size: " << lhsShape[1]
-           << "\nPipelinedepth: " << softwarePipelinedepth;
+  // Number of warps in matmul problem dimension M, N, and K.
+  // Note the following thread dim -> problem dim mapping:
+  // DimY -> ProblemDimM, DimX -> ProblemDimN, DimZ -> ProblemDimK.
+  SmallVector<int64_t> numWarps{workgroupSize[kDimY],
```

Comment:
having the num warps be in different order than workgroup size is a confusing. Number of warps should be using X, Y, Z not M, N, K order

Reply:
This is the convention and mapping I followed and drew so, we don't get confuse what threads, block, grid dims are mapped to which dimensions in the problem.

<img width="1072" alt="Screenshot 2023-03-22 at 5 48 05 PM" src="https://user-images.githubusercontent.com/3487276/227070613-149c29b0-dab6-4c07-b67b-0cda59e1a66e.png">

At some level we need to map (x,y,z) to the problem and follow the convention. Using the picture, I followed problem (M,N,K) mapping for everything below the dotted line:

`grid = {bx, by, bz}`
`block/workGroup = {tx, ty, tz}`
--------------------- map architecture's (x,y,z) dims to matmul problem (m, n, k) -------------------------------
`numWarp = {num of warps in [kM], num of warps  in [kN], num of warps in  in [kK]} =  {ty, tx/32, tz}`

`threadBlockShape = {num of tiles in [kM], num of tiles in [kN], num of tiles in [kK]}`

// having the dotted line above numWarp a.k.a. applying the mapping from x,y,z arch dim to m,n,k problem allows nice use of kM for both numerator and denominator in the formula below. The mapping part is happening in `numWarp`
`warpShape = {threadBlockShape[kM]/numWarp[kM], 
                           threadBlockShape[kN]/numWarp[kN],
                           threadBlockShape[kK]/numWarp[kK]}` 
                           
`instructionShape = {num elements in M, num elements in N, num elements in N}` // using m,n,k here is straight forward as it maps nicely to the use of m,n,k in mma.sync*m16n8k16.


Also, I want to mention that to completely avoid this confusion would be to launch threads in just x-dim and create a layout of threads. It maybe out of scope for now, but to support all different layouts say row-row, row-col, col-row, and row-col just having threads in one dim and creating explicit layout gives more control. It also allows having different thread layout for different part of the problem in the same kernel. 

It is just a choice and whatever is easier visualize, we can move the dotted line below numWarp and apply mapping when computing `warpShape`.

Reply:
> It is just a choice and whatever is easier visualize, we can move the dotted line below numWarp and apply mapping when computing warpShape.

That would feel more natural to me as having a number of warp in M, N, K doesn't make sense to me. The number of warp is analogue to number of thread. warpShape feels like the right place to handle the logic.